src/dh-exec.1
.TH "DH\-EXEC" "1" "2023-10-30" "" "dh-exec"
.ad l
.nh
.SH "NAME"
dh\-exec \- Debhelper executable file helpers
.SH "SYNOPSIS"
\fB#! /usr/bin/dh\-exec\fR
.br
src/libfoo-*.so.* debian/foo-plugins/usr/lib/foo/${DEB_HOST_MULTIARCH}/
.br
etc/example.conf => debian/foo/etc/foo/foo.conf
.br
[linux-any kfreebsd-any] some-arch-specific-file /usr/lib/foo/

.SH "DESCRIPTION"
dh\-exec is a simple program, meant to be used as the interpreter for
executable debhelper config files.

It is a wrapper around the various other sub\-commands (see below),
and will pipe the input file through all of them in turn, using an
ordering that makes most sense in the vast majority of cases.

The order as of now is \fIdh\-exec\-subst\fR gets run first, followed
by \fIdh\-exec\-install\fR, so that variable expansion happens before
files need to be copied.

.SH "FILTERING"

In all dh\-exec handled files, lines can be pre\- or post\-fixed with
a list of architectures for which the rest of the line should apply
to. All architectures and wildcards known by
\fIdpkg\-architecture\fR(1) are recognised, even negated ones.

.SH "ARCHITECTURE"

dh\-exec is built up from three layers: there is the \fBdh-exec\fR
utility, its single entry point, the only thing one will need to call.

Below that, there are the various \fIsub\-commands\fR, such as
\fIdh\-exec\-subst\fR, \fIdh\-exec\-filter\fR, \fIdh\-exec\-install\fR
and \fIdh\-exec\-illiterate\fR, which are thin wrappers around the
various dh\-exec scripts, that make sure they only run those that need
to be run.

And the lowest layer are the various scripts that do the actual work.

One can control which sub\-commands to run, or if even more
granularity is desired, one can limit which scripts shall be run,
too. See below for the options!

.SH "OPTIONS"

.IP "\fB\-\-with\fR=\fIcommand\fR[\fB,\fR\fIcommand\fR ...]" 4
Replace the list of sub\-commands to run the input through with a
custom list (where entries are separated by whitespace or
commas) that includes \fIdh\-exec\-filter\fR and \fIdh\-exec\-strip\fR
at the start and end of the queue respectively.
This option will always replace the existing list with
whatever is specified.

This can be used to explicitly set which sub\-commands to use.

The list \fBmust not\fR include the \fIdh\-exec\-\fR prefix.

Defaults to \fIsubst,install\fR.

.IP "\fB\-\-without\fR=\fIcommand\fR[\fB,\fR\fIcommand\fR ...]" 4
Inversely to the option above, this lists all the sub\-commands which
should not be used. The \fIdh-exec-filter\fR and \fIdh-exec-strip\fR
subcommands cannot be removed using this option. Use the
\fB\-\-no-defaults\fR instead.

The list \fBmust not\fR include the \fIdh\-exec\-\fR prefix.

.IP "\fB\-\-with\-scripts=\fIscript\fR[\fB,\fR\fIscript\fR ...]" 4
Replace the list of scripts to run the input through with a custom
list (where entries are separated by whitespace or commas). This
option will always replace the existing list with whatever is
specified.

This can be used to explicitly specify which scripts to use, limiting
even beyond what the \fB\-\-with\fR option is capable of.

The list \fBmust not\fR include the \fIdh\-exec\-\fR prefix.

By default it is empty, meaning there is no filtering done, and
whatever scripts the sub\-commands find, will be run.

.IP "\fB\-\-no\-defaults" 4
Most packages require the sub\-command list to start with
\fIdh\-exec\-filter\fR and end with \fIdh\-exec\-strip\fR
subcommands. If you really want to, use this option to remove
both these sub\-commands.

The \fI\-\-without\fR option does not remove these two
sub\-commands.

.IP "\fB\-\-no\-act" 4
Do not really do anything, but print the pipeline that would have been
run instead.

.IP "\fB\-\-list" 4
List the available sub\-commands and scripts, grouped by sub\-command.

.IP "\fB\-\-help\fR, \fB\-\-version\fR" 4
Display a short help or the package version, respectively.

.SH "SUB\-COMMANDS"

.TP
.B dh\-exec\-subst
Substitutes various variables (either from the environment, or from
\fIdpkg\-architecture\fR(1)).

.TP
.B dh\-exec\-filter
Filters the input in various ways (architecture restrictions, build
profiles, and so on).

.TP
.B dh\-exec\-install
An extension to \fIdh_install\fR(1), that supports renaming files
during the copy process, using a special syntax.

.TP
.B dh\-exec\-strip
Removes unrequired lines such as comments or blank lines.

.SH "ENVIRONMENT"

.PP
\fIDH_EXEC_LIBDIR\fR
.RS 4
The directory in which the wrapped sub\-commands reside. Defaults to
\fI/usr/lib/dh\-exec/\fR.
.RE

.PP
\fIDH_EXEC_SCRIPTDIR\fR
.RS 4
The directory in which the scripts that do the heavy work
live. Defaults to \fI/usr/share/dh\-exec/\fR.
.RE

.SH "FILES"

.PP
\fI$DH_EXEC_LIBDIR/dh\-exec\-*\fR
.RS 4
The various sub\-commands.
.RE

.PP
\fI$DH_EXEC_SCRIPTDIR/dh\-exec\-*\fR
.RS 4
The various scripts ran by the sub\-commands.
.RE

.SH "SEE ALSO"
\fIdebhelper\fR(1),
\fIdh\-exec\-subst\fR(1),
\fIdh\-exec\-filter\fR(1),
\fIdh\-exec\-install\fR(1)

.SH "AUTHOR"
dh\-exec is copyright \(co 2011-2023 by Software Freedom Conservancy, Inc.


debian/changelog
dh-exec (0.28) unstable; urgency=medium

  [Debian Janitor]
  * Remove constraints unnecessary since buster (oldstable):
    + Build-Depends: Drop versioned constraint on perl:any (>= 5.14.2~).
    + Build-Depends: Drop versioned constraint on libdpkg-perl (>= 1.17.14~).

  [Craig Small]
  * install-rename: Look in debian/tmp for files to move Closes: #1028347
  * dh-exec: Add no-defaults to not run filter and strip.
    Closes: #840682
  * Update to standards 4.6.2, no changes required.

 -- Craig Small <csmall@debian.org>  Mon, 30 Oct 2023 17:33:45 +1100

dh-exec (0.27) unstable; urgency=medium

  * install-rename: Remove noop for -i -a flags Closes: #1017023

 -- Craig Small <csmall@debian.org>  Mon, 15 Aug 2022 20:10:45 +1000

dh-exec (0.26) unstable; urgency=medium

  * Replace the QA uploads and put them into git

 -- Craig Small <csmall@debian.org>  Sat, 23 Jul 2022 22:01:49 +1000

dh-exec (0.25) unstable; urgency=medium

  [ Debian Janitor ]
  * Use secure copyright file specification URI.
  * Bump debhelper from old 9 to 12.
  * Change priority extra to priority optional.
  * Drop unnecessary dependency on dh-autoconf.

  [ Craig Small ]
  * Adopting package Closes: #851746
  * Update to standards 4.6.1, no changes required.
  * Bump debhelper to 13.
  * Remove old versioned dependencies where not needed.

 -- Craig Small <csmall@debian.org>  Thu, 21 Jul 2022 19:22:59 +1000

dh-exec (0.24) unstable; urgency=medium

  * Add support for moving files to a package, instead of copying. Thanks
    to Barak A. Pearlmutter and Niels Thykier for the report and bringing
    the request to my attention. (Closes: #498067)
  * Fix the combination of dh_install --fail-missing, dh-exec and
    arch-only/arch-indep builds. Thanks to Ferenc Wágner for the report.
    (Closes: #831786)
  * Bump Standards-Version to 3.9.8, no changes required.

 -- Gergely Nagy <algernon@madhouse-project.org>  Fri, 07 Oct 2016 11:30:20 +0200

dh-exec (0.23.4) unstable; urgency=medium

  * QA upload.

  [ Daniel Bungert ]
  * Fix unintended whitespace removal (LP: #1920825)  Per the tests it's
    desired to remove the whitespace following a build profile - such as
    "<A> B" resulting in "B", but leave lines with no build profile alone.

 -- Matthias Klose <doko@debian.org>  Wed, 31 Mar 2021 15:00:30 +0200

dh-exec (0.23.3) unstable; urgency=medium

  * QA upload.

  [ Helmut Grohne ]
  * Fix FTCBFS: (Closes: #984610)
    + Annotate perl dependency :any.
    + Annotate bats dependency <!nocheck>.

 -- Niels Thykier <niels@thykier.net>  Sat, 06 Mar 2021 09:39:39 +0000

dh-exec (0.23.2) unstable; urgency=medium

  * QA upload.
  * Remove redundant Build-Depends on fakeroot as this weakens a
    bootstrapping cycle.
  * Set Rules-Requires-Root to no as dh-exec does not need
    (fake)root to build.
  * Explicitly pass dh --builddirectory instead of implicitly using
    -B (via DH_OPTIONS) as this enables dh to optimize out some
    redundant helpers.
  * Bump priority to optional as extra has been deprecated.

 -- Niels Thykier <niels@thykier.net>  Sat, 18 May 2019 12:44:04 +0000

dh-exec (0.23.1) unstable; urgency=medium

  * Orphaning the package.

 -- Gergely Nagy <algernon@madhouse-project.org>  Thu, 03 Jan 2019 18:00:53 +0100

dh-exec (0.23) unstable; urgency=high

  * Fix the DH_CONFIG_ACT_ON_PACKAGES support so that it works on packages
    with dots in their name too. Thanks to Chris Lamb, Luca Bruno, Niels
    Thykier, and Mattia Rizzolo for the report and help in tracking down
    the issue. (Closes: #809283)
  * Urgency set to high, because the above regression causes other
    packages to FTBFS.

 -- Gergely Nagy <algernon@madhouse-project.org>  Sat, 02 Jan 2016 18:05:04 +0100

dh-exec (0.22) unstable; urgency=low

  * dh-exec-filter(1): Correctly support multiple negated arches. Thanks
    to Paul Gevers and Edmund Grimley Evans for the report and testing.
    (Closes: #801809)
  * dh-exec-install-rename(1): Restore d/install support. Thanks to all
    the people reporting the issue! (Closes: #802034)

 -- Gergely Nagy <algernon@madhouse-project.org>  Wed, 28 Oct 2015 11:30:42 +0100

dh-exec (0.21) unstable; urgency=medium

  * Fix a regression in dh-exec-filter(1), which made multiple arch
    wildcards on the same line fail to parse. Thanks Paul Gevers for the
    report! (Closes: #801716)

 -- Gergely Nagy <algernon@madhouse-project.org>  Wed, 14 Oct 2015 11:23:16 +0200

dh-exec (0.20) unstable; urgency=medium

  * dh-exec-filter(1) will only match architecture wildcards now, and will
    keep shell globs intact. Thanks Evgeni Golov for the report!
    (Closes: #801516)
  * Fix a formatting error on the dh-exec-filter(1) man page.
  * Add a "When not to use dh-exec" section to README.md.

 -- Gergely Nagy <algernon@madhouse-project.org>  Mon, 12 Oct 2015 08:47:35 +0200

dh-exec (0.19) unstable; urgency=medium

  * Fix the DH_CONFIG_ACT_ON_PACKAGES support, so it will work with
    packages with regexp specials in their name, such as g++. Thanks to
    Niels Thykier for the report!
  * Updated the copyrights in the source files too, not only in
    debian/copyright.

 -- Gergely Nagy <algernon@madhouse-project.org>  Tue, 06 Oct 2015 08:40:46 +0200

dh-exec (0.18) unstable; urgency=medium

  * Add support for DH_CONFIG_ACT_ON_PACKAGES, see #698054 for details.
    As a consequence, depend and build-depend on debhelper (>= 9.20151004~).
  * Copyright updated, to reflect copyright being assigned to the SFC.

 -- Gergely Nagy <algernon@madhouse-project.org>  Mon, 05 Oct 2015 11:48:24 +0200

dh-exec (0.17) unstable; urgency=medium

  * Do not error out when build profile stanzas are found, but
    DEB_BUILD_PROFILES is empty: let Dpkg::BuildProfiles handle that
    instead. Thanks Samuel Thibault <sthibault@debian.org> for the report!
    (Closes: #799552)

 -- Gergely Nagy <algernon@madhouse-project.org>  Mon, 21 Sep 2015 09:35:59 +0200

dh-exec (0.16) unstable; urgency=medium


src/dh-exec.c
/* dh-exec.c -- Wrapper around dh-exec-* commands.
 * Copyright (C) 2011-2023  Software Freedom Conservancy, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#define _GNU_SOURCE

#include "config.h"

#include <pipeline.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <dirent.h>
#include <errno.h>
#include <stdbool.h>

#include "dh-exec.lib.h"

const char *DH_EXEC_CMD_PREFIX = "dh-exec-";

#define DH_EXEC_CMD_SEPARATORS ",; \t"

static void
dh_exec_pipeline_add (pipeline *p, const char *cmd)
{
  char *path = dh_exec_cmd_path (dh_exec_libdir (), cmd);
  pipeline_command_args (p, path, NULL);
  free (path);
}

static void
dh_exec_cmdlist_free (char **cmdlist)
{
  int i = 0;

  if (!cmdlist)
    return;

  while (cmdlist[i])
    {
      free (cmdlist[i]);
      i++;
    }
  free (cmdlist);
}

static char **
dh_exec_with (char **cmdlist, const char *prglist)
{
  int i = 0;
  char *t, *orig, *curr;

  orig = strdup (prglist);
  t = orig;
  while (strsep (&t, DH_EXEC_CMD_SEPARATORS))
    i++;
  free (orig);

  orig = strdup (prglist);
  t = orig;

  dh_exec_cmdlist_free (cmdlist);

  cmdlist = (char **)calloc (i + 1, sizeof (char *));
  i = 0;

  while ((curr = strsep (&t, DH_EXEC_CMD_SEPARATORS)) != NULL)
    cmdlist[i++] = strdup (curr);
  free (orig);

  cmdlist[i] = NULL;
  return cmdlist;
}

static char **
dh_exec_without (char **cmdlist, const char *prglist)
{
  char *t, *orig, *prg;

  orig = strdup (prglist);
  t = orig;

  while ((prg = strsep (&t, DH_EXEC_CMD_SEPARATORS)) != NULL)
    {
      int i = 0;

      while (cmdlist[i])
        {
          if (strcmp (cmdlist[i], prg) == 0)
            {
              free (cmdlist[i]);
              cmdlist[i] = strdup ("");
            }
          i++;
        }
    }
  free (orig);

  return cmdlist;
}

static int
dh_exec_help (void)
{
  printf ("dh-exec - Scripts to help with executable debhelper files.\n"
          "\n"
          "Usage: dh-exec [OPTION...] [FILE]\n"
          "\n"
          "  --with=[command,...]        Run with the specified sub-commands only.\n"
          "  --without=[command,...]     Run without the specified sub-commands.\n"
          "  --with-scripts=[script,...] Run with the specified scripts only.\n"
          "  --no-defaults               Do not run the default sub-commands.\n"
          "  --no-act                    Do not run, just print the pipeline.\n"
          "  --list                      List the available sub-commands and scripts.\n"
          "  --help                      Display this help screen.\n"
          "  --version                   Output version information and exit.\n"
          "\n"
          "When listing helpers, the list needs to be comma separated.\n"
          "\n"
          "For complete documentation, see the dh-exec(1) manual page.\n");
  return EXIT_SUCCESS;
}

static int
dh_exec_version (void)
{
  printf ("dh-exec " PACKAGE_VERSION "\n"
          "\n"
          "Copyright (C) 2011-2023 Software Freedom Conservancy, Inc.\n"
          "This is free software; see the source for copying conditions.  There is NO\n"
          "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
          "\n"
          "Written by Gergely Nagy <algernon@debian.org> and Craig Small <csmall@debian.org>\n");
  return EXIT_SUCCESS;
}

static int
dh_exec_list_filter (const struct dirent *entry)
{
  if (strncmp (entry->d_name, DH_EXEC_CMD_PREFIX,
               strlen (DH_EXEC_CMD_PREFIX)) != 0 ||
      strncmp (entry->d_name + strlen (DH_EXEC_CMD_PREFIX), "strip",
               strlen ("strip")) == 0)
    return 0;
  return 1;
}

static int
dh_exec_list (char *argv[])
{
  struct dirent **cmdlist;
  int n;
  int cplen = strlen (DH_EXEC_CMD_PREFIX);

  printf ("dh-exec - Available sub-commands and scripts\n"
          "\n");

  n = scandir (dh_exec_libdir (), &cmdlist, dh_exec_list_filter, alphasort);
  if (n < 0)
    {
      fprintf (stderr, "%s: scandir(\"%s\"): %s\n", argv[0],
               dh_exec_libdir (), strerror (errno));
      return 1;
    }

  while (n--)
    {
      struct dirent **scriptlist;
      int sn;

      printf (" %s:\n", cmdlist[n]->d_name + cplen);

      sn = scandir (dh_exec_scriptdir (), &scriptlist, dh_exec_list_filter,
                    alphasort);
      if (sn < 0)
        {
          fprintf (stderr, "%s: scandir(\"%s\"): %s\n", argv[0],
                   dh_exec_scriptdir (), strerror (errno));
          free (cmdlist);
          return 1;
        }

      while (sn--)
        {
          if (strncmp (scriptlist[sn]->d_name, cmdlist[n]->d_name,
                       strlen (cmdlist[n]->d_name)) != 0)
            {
              free (scriptlist[sn]);
              continue;
            }

          printf (" \t%s\n", scriptlist[sn]->d_name + cplen);
          free (scriptlist[sn]);
        }

      free (cmdlist[n]);
      free (scriptlist);
    }

  free (cmdlist);

  return 0;
}

int
main (int argc, char *argv[])
{
  pipeline *p;
  int status, n = 0, no_act = 0, do_list = 0;
  bool use_defaults = true;
  const char *src;

  char **dhe_commands, **dhe_scripts = NULL;
  static struct option dhe_options[] = {
    {"with",    required_argument, NULL, 'I'},
    {"with-scripts", required_argument, NULL, 'i'},
    {"without", required_argument, NULL, 'X'},
    {"no-defaults", no_argument, NULL, 'D'},
    {"help",    no_argument      , NULL, 'h'},
    {"version", no_argument      , NULL, 'v'},
    {"no-act",  no_argument      , NULL, 'n'},
    {"list",    no_argument      , NULL, 'l'},
    {NULL,      0                , NULL,  0 },
  };

  dhe_commands = dh_exec_with (NULL, "subst,install");

  while (1)
    {
      int option_index, c;

      c = getopt_long (argc, argv, "hI:i:X:vnl", dhe_options, &option_index);
      if (c == -1)
        break;

      switch (c)
        {
        case 'I':
          dhe_commands = dh_exec_with (dhe_commands, optarg);
          break;
        case 'X':
          dhe_commands = dh_exec_without (dhe_commands, optarg);
          break;
        case 'i':
          dhe_scripts = dh_exec_with (dhe_scripts, optarg);
          break;
        case 'h':
          dh_exec_cmdlist_free (dhe_commands);
          dh_exec_cmdlist_free (dhe_scripts);
          return dh_exec_help ();
        case 'v':
          dh_exec_cmdlist_free (dhe_commands);
          dh_exec_cmdlist_free (dhe_scripts);
          return dh_exec_version ();
        case 'n':
          no_act = 1;
          break;
        case 'D':
          use_defaults = false;
          break;
        case 'l':
          do_list = 1;
          break;
        default:
          dh_exec_help ();
          dh_exec_cmdlist_free (dhe_commands);
          dh_exec_cmdlist_free (dhe_scripts);
          return (EXIT_FAILURE);
        }
    }

  if (do_list)
    {
      dh_exec_cmdlist_free (dhe_commands);
      dh_exec_cmdlist_free (dhe_scripts);
      return dh_exec_list (argv);
    }

  src = dh_exec_source (argc, optind, argv);

  p = pipeline_new ();

  if (src)
    {
      pipeline_want_infile (p, src);
      setenv ("DH_EXEC_SOURCE", src, 1);
    }

  if (use_defaults)
    dh_exec_pipeline_add (p, "dh-exec-filter");

  if (dhe_scripts)
    {
      int i = 0;
      int size = 0, pos = 0;
      char *env;

      while (dhe_scripts[i])
        {
          size += strlen (dhe_scripts[i]) + 1;
          i++;
        }

      env = (char *)malloc (size + 1);
      i = 0;

      while (dhe_scripts[i])
        {
          char *path;

          if (asprintf (&path, "%s/dh-exec-%s", dh_exec_scriptdir (),
                        dhe_scripts[i]) <= 0)
            {
              perror ("asprintf");
              exit (1);
            }

          if (access (path, R_OK | X_OK) != 0)
            {
              fprintf (stderr, "%s: script '%s' is not valid!\n", argv[0],
                       dhe_scripts[i]);
              exit (1);
            }
          free (path);

          memcpy (env + pos, dhe_scripts[i], strlen (dhe_scripts[i]));
          pos += strlen (dhe_scripts[i]);
          env[pos++] = '|';
          i++;
        }
      env[pos] = '\0';

      setenv ("DH_EXEC_SCRIPTS", env, 1);

      free (env);
      dh_exec_cmdlist_free (dhe_scripts);
    }

  while (dhe_commands[n])
    {
      char *cmd;

      if (dhe_commands[n][0] == '\0')
        {
          n++;
          continue;
        }

      if (asprintf (&cmd, "%s%s", DH_EXEC_CMD_PREFIX, dhe_commands[n]) <= 0)
        {
          perror ("asprintf");
          exit (EXIT_FAILURE);
        }

      dh_exec_pipeline_add (p, cmd);
      free (cmd);

      n++;
    }

  dh_exec_cmdlist_free (dhe_commands);

  if (use_defaults)
    dh_exec_pipeline_add (p, "dh-exec-strip");

  if (no_act)
    {
      pipeline_dump (p, stdout);
      pipeline_free (p);
      return EXIT_SUCCESS;
    }

  pipeline_start (p);
  status = pipeline_wait (p);
  pipeline_free (p);

  return status;
}
